### PR TITLE
Added code example for var<workgroup> page.

### PR DESCRIPTION
**Pull request by cmhhelgeson**
Added an example showing how to perform parallel reduction on an array of input data by synchronizing updates to a var<workgroup> variable with workgroupBarriers.

### CODE REVIEW

**Comment by cmhhelgeson** [comment]
Added myself to the CLA

File: content/variables/var-workgroup.wgsl [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  //Wait for each invocation to finish their data population
+  workgroupBarrier();
+
+  //Get the sum of the elements in the array
+  //  Input Data:  [0, 1, 2, 3, 4, 5, 6, 7]
+  // Loop Pass 1:   [1, 5, 9, 13, 4, 5, 6, 7]
+  // Loop Pass 2:   [6, 22, 9, 13, 4, 5, 6, 7]
+  for (var currentSize = 4u; currentSize >= 1u; currentSize = currentSize / 2u) {
+    if (local_id.x < currentSize) {
+      workgroup_data[local_id.x] = workgroup_data[local_id.x * 2] + workgroup_data[local_id.x * 2 + 1]
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:26]: I'm using lines 514-520 of this code as reference: https://github.com/0hq/WebGPT/blob/760eaeaf83e7941550f1d42df12853be8847be54/condensed/condensed.js#L497

My understanding is that each invocation goes through the loop once, then waits at the workgroupBarrier within the loop before the loop is continued.

So to break it down as I understand it...

**Loop 1**
Loop 1 Invocation 0: workgroup_data[0] = workgroup_data[0] + workgroup_data[1]
Loop 1 Invocation 1: workgroup_data[1] = workgroup_data[2] + workgroup_data[3]
Loop 1 Invocation 2: workgroup_data[2] = workgroup_data[4] + workgroup_data[5]
Loop 1 Invocation 3: workgroup_data[3] = workgroup_data[5] + workgroup_data[6]
Loop 1 Invocation 4, 5, 6, 7: local_id.x > current_size of 4 so condition is failed

All Invocations wait at the workgroupBarrier within the loop before continuing

**Loop 2**
Loop 1 Invocation 0: workgroup_data[0] = workgroup_data[0] + workgroup_data[1]
Loop 1 Invocation 1: workgroup_data[1] = workgroup_data[2] + workgroup_data[3]
Loop 2 Invocation 2, 3, 4, 5, 6, 7 local_id.x > current_size of 2 so condition is failed

etc...

File: content/variables/var-workgroup.wgsl [hunk 2]
```diff
@@ -0,0 +1,37 @@
+//Create zero-initialized workgroup shared data
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:1]: Fixed

File: content/variables/var-workgroup.wgsl [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+var<workgroup> workgroup_data: array<u32, 8>;
+
+@group(0) @binding(0) var<storage, read> input_data: array<u32>;
+@group(0) @binding(1) var<storage, read_write> output_data: array<u32, 1>;
+
+//Our workgroup will execute 64 invocations of the shader, one
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:7]: Fixed

[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:7]: Replaced with 8

File: content/variables/var-workgroup.wgsl [hunk 4]
```diff
@@ -0,0 +1,37 @@
+//Create zero-initialized workgroup shared data
+var<workgroup> workgroup_data: array<u32, 8>;
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:2]: Added const

File: content/variables/var-workgroup.wgsl [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  workgroup_data[local_id.x] = input_data[local_id.x];
+  
+  //Wait for each invocation to finish their data population
+  workgroupBarrier();
+
+  //Get the sum of the elements in the array
+  //  Input Data:  [0, 1, 2, 3, 4, 5, 6, 7]
+  // Loop Pass 1:   [1, 5, 9, 13, 4, 5, 6, 7]
+  // Loop Pass 2:   [6, 22, 9, 13, 4, 5, 6, 7]
+  for (var currentSize = 4u; currentSize >= 1u; currentSize = currentSize / 2u) {
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:24]: Added const

[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:24]: Removed u suffixes.

File: content/variables/var-workgroup.wgsl [hunk 6]
```diff
@@ -0,0 +1,37 @@
+//Create zero-initialized workgroup shared data
+var<workgroup> workgroup_data: array<u32, 8>;
+
+@group(0) @binding(0) var<storage, read> input_data: array<u32>;
+@group(0) @binding(1) var<storage, read_write> output_data: array<u32, 1>;
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:5]: Replaced array with u32.

File: content/variables/var-workgroup.wgsl (hunk 1 above)
[the remark this answers is not in the document]
Comment (ben-clayton) [reply content/variables/var-workgroup.wgsl:26]: > Loop 1 Invocation 1: workgroup_data[1] = workgroup_data[2] + workgroup_data[3]
> Loop 1 Invocation 2: workgroup_data[2] = workgroup_data[4] + workgroup_data[5]

`workgroup_data[2]` is being both read and written to, concurrently, by two invocations. This is a data race.

[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:26]: Yeah, you're right. Sorry about that, will fix it later tonight.

**Review by ben-clayton — COMMENTED** [review]
Looking much better, thank you. A bunch of suggestions, to improve readability and understanding of the barriers.

File: content/variables/var-workgroup.wgsl [hunk 7]
```diff
@@ -1,37 +1,42 @@
-//Create zero-initialized workgroup shared data
+// Create zero-initialized workgroup shared data
+const workgroup_len = 8;
 var<workgroup> workgroup_data: array<u32, 8>;
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:3]: Fixed

File: content/variables/var-workgroup.wgsl [hunk 8]
```diff
@@ -1,37 +1,42 @@
...
 
 @group(0) @binding(0) var<storage, read> input_data: array<u32>;
-@group(0) @binding(1) var<storage, read_write> output_data: array<u32, 1>;
+@group(0) @binding(1) var<storage, read_write> output_data: u32;
 
-//Our workgroup will execute 64 invocations of the shader, one
-@compute @workgroup_size(8, 1, 1)
+// Our workgroup will execute 8 invocations of the shader
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:8]: Fixed

File: content/variables/var-workgroup.wgsl [hunk 9]
```diff
@@ -1,37 +1,42 @@
...
+    if (local_id.x < current_size) {
+      // Read current values from workgroup_data
+      var sum = workgroup_data[local_id.x * 2] + workgroup_data[local_id.x * 2 + 1];
+      // Wait till all current values are read
+      workgroupBarrier();
+      // Write sum to workgroup_data
+      workgroup_data[local_id.x] = sum;
     }
-    //Wait for each invocation to attempt one iteration of the loop
+    // Wait for each invocation to finish one iteration of the loop
```
[the remark this answers is not in the document]
Comment (cmhhelgeson) [reply content/variables/var-workgroup.wgsl:33]: Fixed

File: content/variables/var-workgroup.wgsl [hunk 10]
```diff
@@ -22,12 +22,14 @@ fn computeMain(
...
-    if (local_id.x < current_size) {
+    var sum: u32 = 0;
+    if (local_id.x < u32(current_size)) {
       // Read current values from workgroup_data
-      var sum = workgroup_data[local_id.x * 2] + workgroup_data[local_id.x * 2 + 1];
-      // Wait until all invocations have finished reading from workgroup_data, and have calculated their respective sums
-      workgroupBarrier();
-      // Write sum to workgroup_data
+      sum = workgroup_data[local_id.x * 2] + workgroup_data[local_id.x * 2 + 1];
+   	}
```
Comment (ben-clayton) [inline content/variables/var-workgroup.wgsl:29]: nit: bad indentation.

Comment (ben-clayton) [reply content/variables/var-workgroup.wgsl:29]: Might just be Github - the code looks fine in the preview.

**Review by ben-clayton — APPROVED** [review]
Looks good. Thank you for all the changes!

**Pull request merged**